# Task 5: SQL-Based Analysis of Product Sales

## 🧱 Step 1: Database Setup and Migration

This section establishes the SQL environment for analysis. The Chinook dataset, originally provided as a `.sqlite` file, is migrated into a MySQL database using Python and SQLAlchemy.

### 🔧 Steps Performed:
- Connected to the SQLite database
- Extracted all tables and data using `sqlite3` and `pandas`
- Created a new MySQL schema (`chinook`)
- Migrated each table into MySQL using `to_sql()`

This setup ensures that all subsequent queries are executed against a structured relational database in MySQL.


> ⚠️ **Note:** The database connection string below uses placeholder credentials such as `your_username` and `your_password`.  
> Before running this notebook, make sure to replace them with your own MySQL login details.  
> Example: `mysql+pymysql://your_username:your_password@localhost:3306/chinook`

In [ ]:
import sqlite3
import pandas as pd
from sqlalchemy import create_engine

# Step 1: Connect to SQLite
sqlite_conn = sqlite3.connect("Chinook_Sqlite.sqlite")  

# Step 2: Connect to MySQL
# Format: mysql+pymysql://username:password@host:port/database
mysql_engine = create_engine("mysql+pymysql://your_username:your_password@localhost:3306/chinook", echo=True)

# Step 3: Get all table names from SQLite
sqlite_cursor = sqlite_conn.cursor()
sqlite_cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = [row[0] for row in sqlite_cursor.fetchall()]

# Step 4: Migrate each table
for table in tables:
    df = pd.read_sql_query(f"SELECT * FROM {table}", sqlite_conn)
    df.to_sql(table, mysql_engine, if_exists='replace', index=False)
    print(f"✅ Migrated table: {table}")

# Step 5: Close connections
sqlite_conn.close()


In [ ]:
engine = create_engine("mysql+pymysql://your_username:your_password@localhost:3306/chinook")

## 📊 Step 2: Top-Selling Products Analysis

This section identifies the top-selling products by total units sold and revenue.

### 🔗 Why Use a JOIN?

Product details (`Track.Name`) and sales data (`InvoiceLine.Quantity`, `UnitPrice`) are stored in separate tables.  
To analyze product performance, we join `InvoiceLine` with `Track` using `TrackId`.

### 🧠 Business Logic:
- Group by product name
- Sum quantity and revenue
- Sort by revenue descending


In [11]:
query = """
SELECT 
    t.Name AS Product,
    SUM(il.Quantity) AS UnitsSold,
    SUM(il.UnitPrice * il.Quantity) AS Revenue
FROM InvoiceLine il
JOIN Track t ON il.TrackId = t.TrackId
GROUP BY t.Name
ORDER BY Revenue DESC
LIMIT 10;
"""

df_top_products = pd.read_sql(query, engine)
df_top_products


Product  UnitsSold  Revenue
0         The Trooper        5.0     4.95
1  Dazed and Confused        5.0     4.95
2               Pilot        2.0     3.98
3      Gay Witch Hunt        2.0     3.98
4           Walkabout        2.0     3.98
5            Hot Girl        2.0     3.98
6      The Woman King        2.0     3.98
7             The Fix        2.0     3.98
8      Branch Closing        2.0     3.98
9   Phyllis's Wedding        2.0     3.98

### 📊 Interpretation: Top-Selling Products

The query returns the top-selling products based on total units sold and revenue. Key observations:

- **Most Sold Products**:  
  *The Trooper* and *Dazed and Confused* lead with **5 units sold** each, generating **$4.95** in revenue per product.

- **Revenue Distribution**:  
  All other listed products sold **2 units** each, generating **$3.98** in revenue. This suggests a uniform pricing model across many tracks.

- **Pricing Insight**:  
  The revenue per unit appears to be approximately **$0.99**, indicating consistent unit pricing across products.

- **Sales Concentration**:  
  No product shows significantly higher sales volume or revenue, suggesting a relatively flat distribution among top performers.

This result may reflect either limited transaction volume or a dataset focused on individual track purchases rather than bundled sales.


## 🌍 Step 3: Revenue by Region

This step analyzes total revenue generated across different geographic regions.

### 🔗 Tables Involved:
- `Invoice`: contains billing location (`BillingCity`, `BillingCountry`)
- `InvoiceLine`: contains sales details (`UnitPrice`, `Quantity`)

### 🧠 Business Logic:
- Join `InvoiceLine` with `Invoice` using `InvoiceId`
- Group by `BillingCountry` or `BillingCity`
- Sum total revenue per region


In [ ]:
query = """
SELECT 
    i.BillingCountry AS Country,
    SUM(il.UnitPrice * il.Quantity) AS TotalRevenue
FROM InvoiceLine il
JOIN Invoice i ON il.InvoiceId = i.InvoiceId
GROUP BY i.BillingCountry
ORDER BY TotalRevenue DESC;
"""

df_revenue_by_country = pd.read_sql(query, engine)
df_revenue_by_country


Country  TotalRevenue
0              USA        523.06
1           Canada        303.96
2           France        195.10
3           Brazil        190.10
4          Germany        156.48
5   United Kingdom        112.86
6   Czech Republic         90.24
7         Portugal         77.24
8            India         75.26
9            Chile         46.62
10         Ireland         45.62
11         Hungary         45.62
12         Austria         42.62
13         Finland         41.62
14     Netherlands         40.62
15          Norway         39.62
16          Sweden         38.62
17         Belgium         37.62
18       Australia         37.62
19           Spain         37.62
20         Denmark         37.62
21           Italy         37.62
22          Poland         37.62
23       Argentina         37.62

### 🌍 Interpretation: Revenue by Country

This query aggregates total revenue by billing country. Key observations:

- **Top Markets**:  
  The **USA** leads with over **$523.06**, followed by **Canada** and **France**, indicating strong sales concentration in North America and Western Europe.

- **Emerging Regions**:  
  Countries like **Brazil**, **India**, and **Chile** show notable revenue figures, suggesting active markets outside traditional Western regions.

- **Revenue Distribution**:  
  The drop-off from top countries to mid-tier ones (e.g., Czech Republic, Portugal) is gradual, indicating a relatively broad international customer base.

- **Business Implication**:  
  These insights can inform regional marketing strategies, pricing models, and localization efforts for future product launches.


## 📈 Step 4: Monthly Sales Performance

This section analyzes total revenue trends over time by aggregating sales per calendar month.

### 🔗 Why Use a Date Function?

The `InvoiceDate` field contains full timestamps.  
To analyze monthly performance, we extract the **year-month** portion using `DATE_FORMAT(i.InvoiceDate, '%%Y-%%m')`, which is compatible with MySQL.

### 🧠 Business Logic:
- Format `InvoiceDate` to year-month using `DATE_FORMAT`
- Group by formatted month
- Sum total revenue (`Invoice.Total`)
- Sort chronologically to observe trends


In [16]:
query = r"""
SELECT 
    DATE_FORMAT(i.InvoiceDate, '%%Y-%%m') AS Month,
    ROUND(SUM(i.Total), 2) AS MonthlyRevenue
FROM Invoice i
GROUP BY Month
ORDER BY Month;
"""

df_monthly_revenue = pd.read_sql(query, engine)
df_monthly_revenue


Month  MonthlyRevenue
0   2009-01           35.64
1   2009-02           37.62
2   2009-03           37.62
3   2009-04           37.62
4   2009-05           37.62
5   2009-06           37.62
6   2009-07           37.62
7   2009-08           37.62
8   2009-09           37.62
9   2009-10           37.62
10  2009-11           37.62
11  2009-12           37.62
12  2010-01           52.62
13  2010-02           46.62
14  2010-03           44.62
15  2010-04           37.62
16  2010-05           37.62
17  2010-06           37.62
18  2010-07           37.62
19  2010-08           37.62
20  2010-09           36.63
21  2010-10           37.62
22  2010-11           37.62
23  2010-12           37.62
24  2011-01           37.62
25  2011-02           37.62
26  2011-03           37.62
27  2011-04           51.62
28  2011-05           42.62
29  2011-06           50.62
30  2011-07           37.62
31  2011-08           37.62
32  2011-09           37.62
33  2011-10           37.62
34  2011-11           23.76
35  2011-12           37.62
36  2012-01           37.62
37  2012-02           37.62
38  2012-03           37.62
39  2012-04           37.62
40  2012-05           37.62
41  2012-06           37.62
42  2012-07           39.62
43  2012-08           47.62
44  2012-09           46.71
45  2012-10           42.62
46  2012-11           37.62
47  2012-12           37.62
48  2013-01           37.62
49  2013-02           27.72
50  2013-03           37.62
51  2013-04           33.66
52  2013-05           37.62
53  2013-06           37.62
54  2013-07           37.62
55  2013-08           37.62
56  2013-09           37.62
57  2013-10           37.62
58  2013-11           49.62
59  2013-12           38.62

### 📊 Interpretation: Monthly Revenue Trends (2009–2013)

This query reveals monthly revenue patterns over a five-year period.

#### 📈 Observations:

- **Stable Baseline**: Most months report consistent revenue around **37.62**, indicating a steady stream of sales.
- **Seasonal Peaks**:
  - **Jan 2010**: Revenue jumps to **46.62**
  - **Aug 2011**: Highest spike at **50.60**
  - **Aug–Oct 2012**: Sustained increase, peaking at **47.62** and **46.71**
  - **Nov 2013**: Another peak at **49.62**
- **Notable Dips**:
  - **Feb 2013**: Drops to **27.72**, the lowest in the dataset.
  - **Apr 2013**: Slight dip to **33.66**

#### 📊 Business Implications:

- **Marketing Timing**: Revenue spikes suggest successful campaigns or seasonal demand in **January**, **August**, and **November**.
- **Forecasting**: The consistent baseline allows for reliable projections, while peaks and dips highlight months needing deeper analysis.
- **Anomaly Detection**: Sudden drops (e.g., Feb 2013) may indicate operational issues, market shifts, or data anomalies worth investigating.


## 🪟 Step 5: Window Function – Top Product per Artist

This section uses a window function to identify the highest-revenue product for each artist.

### 🔗 Why Use a Window Function?

Standard aggregation (`GROUP BY`) can summarize totals, but it can't rank items within groups.  
To find the top-selling product **per artist**, we use `ROW_NUMBER()` with `PARTITION BY ar.Name`.

### 🧠 Business Logic:
- Join `InvoiceLine`, `Track`, `Album`, and `Artist` to access product and artist info
- Group by artist and product name
- Sum revenue per product
- Use `ROW_NUMBER()` to rank products within each artist group
- Filter or highlight `ProductRank = 1` to get each artist’s top performer


In [18]:
query = """
SELECT 
    ar.Name AS Artist,
    t.Name AS Product,
    ROUND(SUM(il.UnitPrice * il.Quantity), 2) AS Revenue,
    ROW_NUMBER() OVER (PARTITION BY ar.Name ORDER BY SUM(il.UnitPrice * il.Quantity) DESC) AS ProductRank
FROM InvoiceLine il
JOIN Track t ON il.TrackId = t.TrackId
JOIN Album al ON t.AlbumId = al.AlbumId
JOIN Artist ar ON al.ArtistId = ar.ArtistId
GROUP BY ar.Name, t.Name
ORDER BY ar.Name, ProductRank;
"""

df_ranked_products = pd.read_sql(query, engine)
df_ranked_products


Artist                                            Product  \
0              AC/DC                                   Inject The Venom   
1              AC/DC                                           Overdose   
2              AC/DC                                         Snowballed   
3              AC/DC                                 Breaking The Rules   
4              AC/DC                              Put The Finger On You   
...              ...                                                ...   
1911  Zeca Pagodinho                        Sem Essa de Malandro Agulha   
1912  Zeca Pagodinho                               Coração Em Desalinho   
1913  Zeca Pagodinho                                      Faixa Amarela   
1914  Zeca Pagodinho                                            Verdade   
1915  Zeca Pagodinho  Feirinha da Pavuna/Luz do Repente/Bagaço da La...   

      Revenue  ProductRank  
0        1.98            1  
1        1.98            2  
2        1.98            3  
3        0.99            4  
4        0.99            5  
...       ...          ...  
1911     0.99            5  
1912     0.99            6  
1913     0.99            7  
1914     0.99            8  
1915     0.99            9  

[1916 rows x 4 columns]

### 🧮 Interpretation: Top Product per Artist Using ROW_NUMBER()

This query ranks each artist’s products by total revenue, allowing us to identify their top-selling track.

#### 🎸 Example: AC/DC
- **Top Track**: "Inject The Venom" with **$1.98** in revenue
- Other high performers: "Overdose" and "Snowballed" also earned **$1.98**, suggesting equal popularity
- Lower-ranked tracks like "Breaking The Rules" and "Put The Finger On You" earned **$0.99**

#### 🎤 Example: Zeca Pagodinho
- All listed tracks earned **$0.99**, indicating uniform sales across his catalog
- Top-ranked track: "Sem Essa de Malandro Agulha"
- Ranking helps distinguish performance even when revenue is tied

#### 📊 Business Implication:
- **ProductRank = 1** isolates each artist’s best-performing track
- Useful for curating featured content, optimizing playlists, or targeting promotional efforts
- Highlights catalog depth and revenue concentration per artist
